src/github.com/matrix-org/dendrite/clientapi/writers/joinroom.go

```diff
@@ -0,0 +1,249 @@
+
+	needed, err := gomatrixserverlib.StateNeededForEventBuilder(&eb)
+	if err != nil {
+		return httputil.LogThenError(r.req, err)
+	}
+
+	// Ask the roomserver for information about this room
+	queryReq := api.QueryLatestEventsAndStateRequest{
+		RoomID:       roomID,
+		StateToFetch: needed.Tuples(),
+	}
+	var queryRes api.QueryLatestEventsAndStateResponse
+	if queryErr := r.queryAPI.QueryLatestEventsAndState(&queryReq, &queryRes); queryErr != nil {
+		return httputil.LogThenError(r.req, queryErr)
+	}
+
+	if queryRes.RoomExists {
+		// TODO: Implement joining rooms that already the server is already in.
+		// This should just fall through to the usual event sending code.
+		panic(fmt.Errorf("Joining rooms that the server already in is not implemented"))
+	}
+
+	var event gomatrixserverlib.Event
+	var respMakeJoin gomatrixserverlib.RespMakeJoin
+	var respSendJoin gomatrixserverlib.RespSendJoin
+	for _, server := range servers {
+		respMakeJoin, err = r.federation.MakeJoin(server, roomID, r.userID)
+		if err != nil {
+			// TODO: Check if the user was not allowed to join the room.
+
+			// Try the next server in the list.
+			continue
+		}
+
+		// Set all the fields to be what they should be, this should be a no-op
+		// but it's possible that the remote server returned us something "odd"
+		r.writeToBuilder(&respMakeJoin.JoinEvent, roomID)
+
+		now := time.Now()
+		eventID := fmt.Sprintf("$%s:%s", util.RandomString(16), r.cfg.ServerName)
+		if event, err = respMakeJoin.JoinEvent.Build(
+			eventID, now, r.cfg.ServerName, r.cfg.KeyID, r.cfg.PrivateKey,
+		); err != nil {
+			return httputil.LogThenError(r.req, err)
+		}
+		respSendJoin, err = r.federation.SendJoin(server, event)
+		if err != nil {
+			// Try the next server in the list.
+			continue
+		}
+
+		if err = respSendJoin.Check(r.keyRing, event); err != nil {
+			// Try the next server in the list.
+			continue
+		}
+
+		if err = r.producer.SendEventWithState(
+			gomatrixserverlib.RespState(respSendJoin), event,
+		); err != nil {
+			return httputil.LogThenError(r.req, err)
+		}
+
+		return util.JSONResponse{
+			Code: 200,
+			// TODO: Put the response struct somewhere common.
+			JSON: struct {
+				RoomID string `json:"room_id"`
+			}{roomID},
+		}
+	}
+
+	if err != nil {
+		// TODO: Gerenerate the correct HTTP status code for all different
+		// kinds of errors that could have happened.
+		// The possible errors include:
+		//   1) We can't connect to the remote servers.
+		//   2) None of the servers we could connect to think we are allowed
+		//	    to join the room.
+		//   3) The remote server returned something invalid.
+		//   4) We couldn't fetch the public keys needed to verify the
+		//      signatures on the state events.
+		//   5) ...
+		return httputil.LogThenError(r.req, err)
+	}
+
+	// This should only be reached if the list of candidate servers was empty.
+	// If the list was non-empty then either the loop body would have returned
+	// or err would be non-nil.
```

Comment:
Maybe put the current for loop in a new function and return the last error on failure?